neural_compressor/compression/pruner/pruning.py

```diff
@@ -0,0 +1,368 @@
+"""Pruning."""
+# !/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2023 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from neural_compressor.compression.pruner.utils import process_config, parse_to_prune, get_sparsity_ratio
+from neural_compressor.compression.pruner.pruners import get_pruner
+from neural_compressor.compression.pruner.utils import logger, torch, collect_layer_inputs, get_layers
+from typing import Optional
+from tqdm.auto import tqdm
+PRUNINGS = {}
+
+def register_pruning(name):
+    """Class decorator to register a pruning subclass to the registry.
+
+    Decorator function used before a pruner subclass.
+    Make sure that the pruning class decorated by this function can be registered in PRUNINGS.
+
+    Args:
+        cls (class): The subclass of register.
+        name: A string. Define the pruning type.
+
+    Returns:
+        cls: The class of register.
+    """
+
+    def register(pruning):
+        PRUNINGS[name] = pruning
+        return pruning
+
+    return register
+
+
+class BasePruning:
+    """Pruning.
+
+    The main class to do pruning; it contains at least one Pruner object.
+
+    Args:
+        config: a string representing the path to a config file. For config file template, please refer to
+            https://github.com/intel/neural-compressor/tree/master/examples/pytorch/nlp/huggingface_models/text-classification/pruning/pytorch_pruner/eager/
+
+    Attributes:
+        model: The model object to prune.
+        config_file_path: A string representing the path to a config file.
+        pruners: A list. A list of Pruner objects.
+        pruner_info: A config dict object that contains pruners' information.
+    """
+    
+    def __init__(self, config, model: torch.nn.Module):
+        """Initialize."""
+        self._model = model
+        self.pruners_info = config
+        self.pruners = self._generate_pruners(model)
+
+    def _generate_pruners(self, model: torch.nn.Module):
+        """Generate pruners.
+
+        :param config: WeightPruningConfig
+        :param model: The torch module to be pruned
+        :return: A list of pruner
```

Comment:
need to support inc prepare_compression later, I don't see  any change for this part